### PR TITLE
 [JENKINS-45657] Fix & Enhancement to filter branches

### PR DESCRIPTION
**Pull request by fbyrne**
[JENKINS-45657] Fix to use script path defined in Multi Branch Pipeline
[JENKINS-TO-BE-CREATED] Enhancement to provide branch regex to deal with large number of branches.

### CODE REVIEW

**Comment by p4paul** [comment]
Hi Fergus, 

Thank you for the Pull Request.  I have been giving this area some thought as the feature is relatively new and not all the use cases have been captured.  

The BranchesScmSource use `p4 dirs` to find branch candidates from the list of`Include branches` then looks for a `Jenkinsfile` at the root of each branch.  You could use a `p4 dirs` search pattern in the `Include branches` list (e.g. `//depot/*/components`).  Currently the plugin strips any trailing `...` and adds a `/*` to the end of the path, which may not be that helpful (see cleanDirPaths).  The directory name at the end of each path then becomes the branch name.


e.g. given a p4 dirs of `//depot/projA/*` you get two branches `main` and `r1.0`:
```
//depot/projA/main/ --> 'main'
//depot/projA/r1.0/ --> 'r1.0'
```

This may not suit all users as the last directory name may not be a suitable branch name. 

e.g. a p4 dirs of `//depot/projA/*/components/*` would clash with the branch names:
```
//depot/projA/main/components/a --> 'a'
//depot/projA/r1.0/components/a --> 'a'
//depot/projA/main/components/b --> 'b'
//depot/projA/r1.0/components/b --> 'b'
```

Perhaps a group matcher with a regex?

e.g. `//depot/projA/(.*)/components/(.*)` and a name builder `{$2}-{$1}` giving:
```
//depot/projA/main/components/a --> 'a-main'
//depot/projA/r1.0/components/a --> 'a-r1.0'
//depot/projA/main/components/b --> 'b-main'
//depot/projA/r1.0/components/b --> 'b-r1.0'
```

In addition the subsequent search for a `Jenkinsfile` may not suit all users as some change the name or store it in a non-root locations (e.g. scripts/).

**Comment by p4paul** [comment]
Happy to merge in the changes, but please can you update the following:
- [ ] BranchesScmSource: change the display name back to `Helix Branches` as Source seemed a bit vague. 
- [ ] Please revert P4EnvironmentContributor and PerforceScmSourceTest as there are no code changes just white space.

File: src/main/java/org/jenkinsci/plugins/p4/scm/BranchesScmSource.java [hunk 1]
```diff
@@ -1,88 +1,140 @@
...
+		return defaultScriptPath;
+	}
+
+	@Extension
+	@Symbol("multiBranch")
+	public static final class DescriptorImpl extends P4ScmSourceDescriptor {
+
+		@Override
+		public String getDisplayName() {
+			return "Helix Source";
```
Comment (fbyrne) [inline src/main/java/org/jenkinsci/plugins/p4/scm/BranchesScmSource.java]: Will change back

**Comment by fbyrne** [comment]
Closing after updating with another [PR 59](https://github.com/jenkinsci/p4-plugin/pull/59)

**Pull request closed without merging**